### PR TITLE
Issue-4078: Data Provider - Parallelize tests with multiple data sets

### PR DESCRIPTION
**Pull request by yamkay**
## Proposed changes

#4078 - Data Provider - Parallelize tests with multiple data sets

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

Due to issues while rewriting the history of commits, I closed the PR #4096 and opened a new PR #4452

### CODE REVIEW

File: docs/ConfigurationFile.md [hunk 1]
```diff
@@ -62,6 +62,15 @@ exports.config = {
     ],
     //
     // ============
+    // DataProvider
```
Comment (CrispusDH) [inline docs/ConfigurationFile.md]: Would it works only for mocha runner?

Comment (yamkay) [reply docs/ConfigurationFile.md]: It can work with all runners.

**Comment by yamkay** [comment]
Due to issues while rewriting the history of commits, I closed the PR #4096 and opened a new PR #4452

**Comment by martinfrancois** [comment]
@yamkay can we close this PR then?

File: packages/wdio-cli/src/launcher.js [hunk 2]
```diff
@@ -53,6 +53,27 @@ class Launcher {
...
+    getSpecsWithDataProviderAndRetries(config, capSpecs, capExclude) {
+        /**
+         * avoid retries in watch mode
+         */
+        const specFileRetries = config.watch ? 0 : config.specFileRetries
+        let specs = []
+        const dataProvidersMap = this.configParser.getDataProviders() || {}
+
+        this.configParser.getSpecs(capSpecs, capExclude).forEach((specFile) => {
+            let specDataSet = dataProvidersMap[specFile] || ['']
```
Comment (baruchvlz) [inline packages/wdio-cli/src/launcher.js]: Is `['']` needed or can we only run the second loop if a data set is found?

File: packages/wdio-cli/src/launcher.js [hunk 3]
```diff
@@ -53,6 +53,27 @@ class Launcher {
...
+            specDataSet.forEach((data) => {
+                specs.push({ files: [specFile], testData: data, retries: specFileRetries })
+            })
+        })
```
Comment (baruchvlz) [inline packages/wdio-cli/src/launcher.js]: It seems like nothing really happens in the loop if `dataProvidersMap` is an empty object. Can't we return an empty `specs` array if `this.configParser.getDataProviders()` is null?

Example:

```js
const dataProvidersMap = this.configParser.getDataProviders() || {}

if (dataProvidersMap) {
  // do loop here
}

return specs

```

File: packages/wdio-cli/src/launcher.js [hunk 4]
```diff
@@ -361,12 +380,13 @@ class Launcher {
...
      * @param  {Number} retries   Number or retries remaining
+     * @param  {Object} testData  Test data for the instance
      */
-    endHandler ({ cid, exitCode, specs, retries }) {
+    endHandler ({ cid, exitCode, specs, retries, testData }) {
         const passed = exitCode === 0
 
         if (!passed && retries > 0) {
-            this.schedule[parseInt(cid)].specs.push({ files: specs, retries: retries - 1, rid: cid })
+            this.schedule[parseInt(cid)].specs.push({ files: specs, testData: testData, retries: retries - 1, rid: cid })
```
Comment (baruchvlz) [inline packages/wdio-cli/src/launcher.js]: No need to do `testData: testData`, please take advantage of object literals

File: packages/wdio-config/src/dataproviders.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Loads the data providers from the files to memory
+ * @param  {Object} dataProviderFilesPaths  List of files with data provider injections
+ */
+export function initializeDataProviders (dataProviderFilesPaths) {
+    const dataProviderFiles =  ConfigParser.getFilePaths(dataProviderFilesPaths)
+    const dataProvidersMap = {}
+    let currentDataProviderFile = ''
+
+    global.dataProvider = function (specFilePath, dataSet) {
```
Comment (baruchvlz) [inline packages/wdio-config/src/dataproviders.js]: Don't understand why you're attaching this to the global object and then deleting it later in the block. Does it have to be like this?

File: packages/wdio-config/src/dataproviders.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        if (!Array.isArray(dataSet)) {
+            throw new Error(`The data set passed from the file "${currentDataProviderFile}" for the spec file path "${specFilePath}" is not an array.`)
+        }
+
+        let specFiles =  ConfigParser.getFilePaths(specFilePath)
```
Comment (baruchvlz) [inline packages/wdio-config/src/dataproviders.js]: Use const

File: packages/wdio-config/src/dataproviders.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        specFiles.forEach((specFile) => {
+            const specFileDataProvider = dataProvidersMap[specFile]
+            if (typeof specFileDataProvider === 'undefined') {
+                const dataProviderObject = {
+                    'specFilePath': specFilePath,
```
Comment (baruchvlz) [inline packages/wdio-config/src/dataproviders.js]: No need for quotation marks in the object property value, by removing this you can also use object literals to set this data.

**Comment by christian-bromann** [comment]
Closing this as it seems to be no activity on this PR anymore. Feel free to resubmit this patch if there is still an interest.

**Comment by Trueside** [comment]
> Closing this as it seems to be no activity on this PR anymore. Feel free to resubmit this patch if there is still an interest.

Hello! I hope you are doing well. Can you tell me, please, about data-driven testing in wdIO: is there something like this?
I want to run test in parallel with different data suites, like 

```
describe(' TEST ', async () => {
  it(' login ', async () => {
    //some test
  })
})

dataObject = {
  case1: {
    username: 'validName',
    password: 'validPass'
  },
  case2: {
    username: 'invalidName',
    password: 'invalidPass'
  }
}
```

It was very useful in codeception  framework in PHP, I will be very grateful if you can suggest something like this in WebdriverIO!

**Comment by christian-bromann** [comment]
@Trueside you can always do things like this:
```js
describe(' TEST ', async () => {
    for (const case of Object.values(dataObject)) {
        it(`test with ${case.username}`, () => {
            // case.username
            // case.password
        })
    }
})

dataObject = {
    case1: {
        username: 'validName',
        password: 'validPass'
    },
    case2: {
        username: 'invalidName',
        password: 'invalidPass'
    }
}
```

**Comment by SRKarumuri** [comment]
I am more interested in using data Provider, do i need to install any packages, what changes do I need to do in wdio.conf in order to cater for dataprovider.

**Comment by christian-bromann** [comment]
@SRKarumuri WebdriverIO doesn't has a specific feature around data provider.

**Comment by SRKarumuri** [comment]
> @SRKarumuri WebdriverIO doesn't have a specific feature around data provider.
Thank you for your response,
I have seen the following video "https://www.youtube.com/watch?v=0YQCVJk8K_Q" from saucelab for dataproviders. is there any wat to run same Test parallelly with different set of Data.

**Comment by christian-bromann** [comment]
The person who have the talk wanted to contribute this capability to the project but never did. If you like to have such a feature become reality, please get involved. I am happy to help when questions arrive.

**Comment by SRKarumuri** [comment]
> The person who have the talk wanted to contribute this capability to the project but never did. If you like to have such a feature become reality, please get involved. I am happy to help when questions arrive.

Sounds like a plan. This feature gives more value and advantage to webdriverio project. If we consider Big domain systems instead of having 100's of test cases which goes through same 20 or 30 pages with different set of data. this feature will add more value where with one test case they can cover all different business scenarios with different set of data. 
Please let me know how I can get involved in this feature.

**Comment by christian-bromann** [comment]
> This feature gives more value and advantage to webdriverio project.

I agree, which is why I was excited that someone would work on this.

> Please let me know how I can get involved in this feature.

I guess first is to create an issue describing how you like to have this Dataprovider work in WebdriverIO. We can comment there and find an interface that we all like. Then I guess a good start are the [contributing docs](https://webdriver.io/docs/contribute). I would start looking into the [launcher](https://github.com/webdriverio/webdriverio/blob/e5f951e242c098f6173dd77802870e5e2dba62e0/packages/wdio-cli/src/launcher.ts) mechanism to see how the scheduler is build. We provide [office hours](https://webdriver.io/community/openofficehours) in case you have questions to the code base and things around the project.

**Comment by SRKarumuri** [comment]
> ```js
> dataObject = {
>     case1: {
>         username: 'validName',
>         password: 'validPass'
>     },
>     case2: {
>         username: 'invalidName',
>         password: 'invalidPass'
>     }
> }
> ```

Will create one ticket tomorrow.

**Comment by SRKarumuri** [comment]
> > This feature gives more value and advantage to webdriverio project.
> 
> I agree, which is why I was excited that someone would work on this.
> 
> > Please let me know how I can get involved in this feature.
> 
> I guess first is to create an issue describing how you like to have this Dataprovider work in WebdriverIO. We can comment there and find an interface that we all like. Then I guess a good start are the [contributing docs](https://webdriver.io/docs/contribute). I would start looking into the [launcher](https://github.com/webdriverio/webdriverio/blob/e5f951e242c098f6173dd77802870e5e2dba62e0/packages/wdio-cli/src/launcher.ts) mechanism to see how the scheduler is build. We provide [office hours](https://webdriver.io/community/openofficehours) in case you have questions to the code base and things around the project.

I have created the new Feature Ticket.
https://github.com/webdriverio/webdriverio/issues/9090

**Comment by Zarmeena** [comment]
> @Trueside you can always do things like this:
> 
> ```js
> describe(' TEST ', async () => {
>     for (const case of Object.values(dataObject)) {
>         it(`test with ${case.username}`, () => {
>             // case.username
>             // case.password
>         })
>     }
> })
> 
> dataObject = {
>     case1: {
>         username: 'validName',
>         password: 'validPass'
>     },
>     case2: {
>         username: 'invalidName',
>         password: 'invalidPass'
>     }
> }
> ```

@christian-bromann How can we execute Selenium tests in parallel through this? It continues to execute with next dataset in the currently focused browser window, itself.

**Pull request closed without merging**